mocks_test.go

```diff
@@ -6,13 +6,30 @@ import (
 )
 
 type mockClient struct {
-	req  *http.Request
+	req        *http.Request
+	resp       http.Response
+	err        error
+	retries    []retry
+	retryIndex int
+}
+
+type retry struct {
 	resp http.Response
 	err  error
 }
 
 func (c *mockClient) Do(req *http.Request) (*http.Response, error) {
 	c.req = req
+	if c.retries != nil {
+		currentRetry := c.retries[c.retryIndex]
```

Comment:
Let's add an explicit panic with meaningful message when index is out of bounds.